### ISSUE TITLE
cnvkit.py batch - no .cns file

### ISSUE THREAD

**Issue by unitsmanolis** [etal/cnvkit#582]
Hi, 

I'm running this code with WGS data for somatic CNV discovery:

source activate cnvkit

```
cnvkit.py batch \
--method wgs \
--segment-method cbs \
tumor.bam \
--normal normal.bam \
--output-reference reference.cnn \
--fasta hg38.fasta \
--target-avg-size 1000 \
--access hg38.bed \
--output-dir results/ \
--diagram --scatter \
--processes 16
```

The point is that does not create a .cns file. I do not have any error or warning. If I take the .cnr file and I go to run manually the followings steps, they works:

```
cnvkit.py segment tumor.cnr -o tumor.cns
cnvkit.py scatter tumor.cnr -s tumor.cns -o tumor-scatter.pdf
cnvkit.py diagram tumor.cnr -s tumor.cns -o tumor-diagram.pdf
```

Could you help me about what to check? 

Many thanks

**Comment by tskir** [comment]
Hi @unitsmanolis! Can you please provide:

1. CNVkit version you are using (`cnvkit.py version`), and
2. The complete log output of the command you're running?

**Comment by unitsmanolis** [comment]
Hi @tskir 

I rerun many time the test changing options ... here one of the log output

--target-avg-size 2500 
--processes 32

1) CNVkit 0.9.7

2) log output

> WGS protocol: recommend '--annotate' option (e.g. refFlat.txt) to help locate genes in output files.
> Detected file format: bed
> Splitting large targets
> Created directory /netapp/test/results
> Wrote /netapp/test/results/hg38_custom_1-22XY.target.bed with 1150191 regions
> Wrote /netapp/test/results/hg38_custom_1-22XY.antitarget.bed with 0 regions
> Building a copy number reference from normal samples...
> Processing reads in normal.bam
> Skip processing normal.bam with empty regions file /netapp/test/results/hg38_custom_1-22XY.antitarget.bed
> Wrote /netapp/test/results/normal.antitargetcoverage.cnn with 0 regions
> Time: 698.345 seconds (951563 reads/sec, 1647 bins/sec)
> Summary: #bins=1150191, #reads=664519039, mean=577.7467, min=0.0, max=337951.3541666667 
> Percent reads in regions: 85.281 (of 779213484 mapped)
> Wrote /netapp/test/results/normal.targetcoverage.cnn with 1150191 regions
> Relative log2 coverage of chrX=-0.98, chrY=-0.823 (maleness=1.6e+03 x 1.38 = 2.21e+03) --> assuming male
> Loading /netapp/test/results/normal.targetcoverage.cnn
> Calculating GC and RepeatMasker content in /shared/Homo_sapiens_assembly38.fasta ...
> Extracting sequences from chromosome chr1
> Extracting sequences from chromosome chr2
> Extracting sequences from chromosome chr3
> Extracting sequences from chromosome chr4
> Extracting sequences from chromosome chr5
> Extracting sequences from chromosome chr6
> Extracting sequences from chromosome chr7
> Extracting sequences from chromosome chr8
> Extracting sequences from chromosome chr9
> Extracting sequences from chromosome chr10
> Extracting sequences from chromosome chr11
> Extracting sequences from chromosome chr12
> Extracting sequences from chromosome chr13
> Extracting sequences from chromosome chr14
> Extracting sequences from chromosome chr15
> Extracting sequences from chromosome chr16
> Extracting sequences from chromosome chr17
> Extracting sequences from chromosome chr18
> Extracting sequences from chromosome chr19
> Extracting sequences from chromosome chr20
> Extracting sequences from chromosome chr21
> Extracting sequences from chromosome chr22
> Extracting sequences from chromosome chrX
> Extracting sequences from chromosome chrY
> Correcting for GC bias...
> Loading /netapp/test/results/normal.antitargetcoverage.cnn
> Calculating average bin coverages
> Calculating bin spreads
> Targets: 18595 (1.6167%) bins failed filters (log2 < -5.0, log2 > 5.0, spread > 1.0)
> Wrote PM12S-reference.cnn with 1150191 regions
> Running 1 samples in 32 processes (that's 32 processes per bam)
> Running the CNVkit pipeline on /netapp04/Manolis_WGS_intermedi/GATKv4130_v1_001_Sara_Polmoni_gDNA_Macrogen_NovaSeq_hg38_wgs_gatk4_1-22_XY/tumor.bam ...
> Processing reads in tumor.bam
> Time: 540.948 seconds (1264780 reads/sec, 2126 bins/sec)
> Summary: #bins=1150191, #reads=684180388, mean=594.8407, min=0.0, max=313646.9375 
> Percent reads in regions: 86.559 (of 790421235 mapped)
> Wrote /netapp/test/results/tumor.targetcoverage.cnn with 1150191 regions
> Skip processing tumor.bam with empty regions file /netapp/test/results/hg38_custom_1-22XY.antitarget.bed
> Wrote /netapp/test/results/tumor.antitargetcoverage.cnn with 0 regions
> Processing target: tumor
> Keeping 1131596 of 1150191 bins
> Correcting for GC bias...
> Processing antitarget: tumor
> Wrote /netapp/test/results/tumor.cnr with 1131596 regions
> Segmenting /netapp/test/results/tumor.cnr ...
> Segmenting with method 'cbs', significance threshold 1e-06, in 32 processes

Many thanks for your time

**Comment by tskir** [comment]
@unitsmanolis That certainly looks very interesting, especially given how segmentation seems to work by itself but not as part of the batch. I have a number of follow-up questions to try and narrow it down:

1. The batch log which you provided stops abruptly right after the segmentation starts. From a user perspective, what actually happens at that point? Does it just terminate without saying anything else, or does it continue to hang indefinitely?

2. In case it terminates, could you check the return code please? This can be done by running the `cnvkit.py batch` command and then, _immediately_ after it returns, checking `echo $?` in bash. It's important that `echo $?` is run just after `cnvkit.py batch` with no other commands in between. Especially I'm wondering if it's zero or non-zero.

3. The only difference I can see as to why segmentation would work by itself, but not as part of the batch, is that by default `cnvkit.py segment` runs in 1 process, while the batch in your case is run with 32 processes, which is passed to the segmentation step. Could you try running `cnvkit.py segment tumor.cnr -o tumor.cns -m cbs -t 0.000001 -p 32` and see if it works or fails?

**Comment by unitsmanolis** [comment]
Hi @tskir

1. The job just finish. I ran it also using SGE (Sun Grid Engine - qsub) and the last line, in all cases, is always the same: "Segmenting with method 'cbs', significance threshold 1e-06, in 32 processes".

2.  The return code is always "0".

3. It works without any error. For example:

$ cnvkit.py segment tumor.cnr -o tumor.cns -m cbs -t 0.000001 -p 16

Dropped 17960 rows with missing log2 values
Segmenting with method 'cbs', significance threshold 1e-06, in 16 processes
Wrote tumor.cns with 2211 regions

I tried to run the code with SGE using several options: "-p 32", "-p 2", "-p 1", "without -p". I ran the "-p 32" code also in my home without SGE and I have always the same problem.

Can I run the "cnvkit.py segment" code manually for all of my samples? just to go on ...

If you need more information/test let me know.

**Comment by unitsmanolis** [comment]
Hi, with the version 0.9.8 works

  source activate cnvkit098
  cnvkit.py batch tumor.bam --normal normal.bam --output-reference reference.cnn \
  --method wgs --segment-method cbs \
  --fasta hg38.fasta --target-avg-size 2500 \
  --access hg38_fixed_1-22_XY.bed \
  --output-dir results/ --diagram --scatter --processes 32

├── reference.cnn
└── results
    ├── hg38_fixed_1-22_XY.antitarget.bed
    ├── hg38_fixed_1-22_XY.target.bed
    ├── normal.antitargetcoverage.cnn
    ├── normal.targetcoverage.cnn
    ├── tumor.antitargetcoverage.cnn
    ├── tumor.bintest.cns
    ├── tumor.call.cns
    ├── tumor.cnr
    ├── tumor.cns
    ├── tumor-diagram.pdf
    ├── tumor-scatter.png
    └── tumor.targetcoverage.cnn

I was expected a scatter pdf file and not png. Is correct .png?

**Comment by unitsmanolis** [comment]
@tskir , do you have any feedback please?
Thanks a lot

**Comment by tskir** [comment]
Hi @unitsmanolis, sorry for disappearing for a while.

First of all, glad you've got it working with a version update! Turns out this was a known issue (https://github.com/etal/cnvkit/issues/491) which was fixed in v0.9.8. This is a proper release (https://github.com/etal/cnvkit/releases/tag/v0.9.8), it's just not displayed on the main page for purely technical reasons, so please feel free to use it.

The scatter format generated by running the `batch` command is indeed always PNG, and unfortunately this is not configurable:
https://github.com/etal/cnvkit/blob/b55ef86c54d499ac34f75675f69bfc20a4afd5e8/cnvlib/batch.py#L215

If you require your scatters to be in the PDF format, I'm afraid you'll have to run the `scatter` command separately to get them.

Please let me know if you have any further questions, I'll be happy to help!

**Comment by unitsmanolis** [comment]
Hi, many thanks for the help!

Have a nice day!

**Comment by AndreaG5** [comment]
Hi everyone,
I followed this issue since I struggle the same. I thought it was about update, but now I don't think so. Previous comment said it was fixed since version 0.9.8, but I am using 0.9.9 and it fails too (same results and outputs of @unitsmanolis ).

P.S. I used same versione (0.9.9) two weeks ago and it worked till the end, doing segmentation and forth. I am wondering what this is about. Any hint?
